google-cloud-storage/lib/google/cloud/storage/bucket.rb

```diff
@@ -1390,6 +1400,14 @@ def create_file file,
           raise ArgumentError, "must provide path" if path.nil?
 
 
+          case checksum
+          when :crc32c
+            raise ArgumentError, "'checksum: :crc32c' should not be provided when a 'crc32c' arg is provided" if crc32c
+            crc32c = File::Verifier.crc32c_for file
+          when :md5
+            raise ArgumentError, "'checksum: :md5' should not be provided when an 'md5' arg is provided" if md5
+            md5 = File::Verifier.md5_for file
+          end
```

Comment:
What will happen if `checksum: :crc32c` and `md5: MY_MD5_VALUE` are both provided? Will it auto-calculate crc32c and then check both?

If checking both is a desired feature, is there a way to `checksum: [:md5, :crc32c]`?

Reply:
I just manually verified that the service will accept both, and if one is incorrect, it will return an error indicating which one. However, if **both** are incorrect, it only returns the error for `crc32c`:

```
Google::Cloud::InvalidArgumentError: invalid: Provided CRC32C "e5jnUQ==" doesn't match calculated CRC32C "Vx8/Nw==".
```

In https://cloud.google.com/storage/docs/hashes-etags, it appears that users are expected to send one or the other, not both:

> When supplying an object's expected MD5 or CRC32C hash in an upload request. Cloud Storage only creates the object if the hash you provide matches the value Cloud Storage calculates.

My take on this is that we shouldn't encourage sending both, since it's probably overkill. One or the other should be enough. But if you have a different read on the docs, I'm happy to consider it.

Reply:
If the goal is to discourage sending both, I'd actually lean toward simply enforcing that you can only do one or the other but not both (e.g. also raise ArgumentError if `checksum: :crc32c` and `md5: MY_MD5_VALUE` are both provided, or vice versa) just to make the behavior clear. It's something that can always be relaxed later.

Reply:
I remembered that the verification for downloads accepts an `:all` value, so I think for consistency we should offer the same here. I'll add it.

Reply:
@andrewsg wrote:

> I'm surprised to see "all" here as we do not support multiple simultaneous checksum types in Python. 

Please see the discussion in this thread regarding multiple simultaneous checksum types. I'll wait to merge until you've had a chance to consider whether `all` is more appropriate than @dazuma's suggestion to prevent accepting both `checksum: :crc32c` and `md5: MY_MD5_VALUE`, for example.

Reply:
Merging now per offline discussion with @dazuma.